### DISCUSSION TITLE
User Identification

### DISCUSSION THREAD

**Discussion by spunkyvt** [dokan-dev/dokan-dotnet#346] [Q&A]
Is there a way to tell which user is accessing the FindFiles when I run this on a server and make the mirror a share?

**Comment by Olof-Lagerkvist** [comment]
Yes, a little bit depending on how authentication to the share is done. But principally, for an AD environment, you get the AD user that the client uses to connect when you check the token from Dokan in this case.

**Reply by spunkyvt** [reply]
I see GetRequestor() on the DokanFileInfo class, but it return an error "Invalid token" when I try to use it.  Is this what you were refering to?

**Reply by Olof-Lagerkvist** [reply]
Okay, I think that running in a debugger and checking what `DokanOpenRequestorToken` actually returns could be useful here. There might be some reason why Dokan driver does not return a token, or there could be something else unexpected happening. Also look for debug messages in debug output, because dokan2.dll prints a debug message with status information in case `DokanOpenRequestorToken` fails.

**Reply by spunkyvt** [reply]
The DokanOpenRequestorToken does return a value, but when combined into the new WindowsIdentity it throws the error invalid token for impersonation- it cannot be duplicated

**Reply by Olof-Lagerkvist** [reply]
Not really sure then, it sounds like the function returns a kind of token that cannot be used for identity in that way then. It would probably need some more investigation. It could be that it needs the token converted using Win32 API first or something like that.

**Reply by spunkyvt** [reply]
So it has to be called in the CreateFile.  It cannot be called anywhere else.  The next hurdle I am facing, I am calling it now and seeing the user, but when I access a share from a workstation as a user, the user being displayed on the server is NT Authority\System and not the actual user
